### PR TITLE
Support for templates in parameters

### PR DESCRIPTION
**Pull request by felix-deschamps**
This document makes a design suggestion for loading parameters from a secondary yaml file

### CODE REVIEW

**Comment by hbuckle** [comment]
This would be incredibly useful to have a single list of environments to pass to templates for multi-stage builds

File: design/parameters.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+# azure-pipelines.yaml
+
+stages: 
+- stage: Group-Stage
+  jobs:
+  - template: groups-template.yaml
+    parameters: 
+    - template: group-data.yaml
+  - template: another-groups-template.yaml
+    paameters:
```
Comment (andywhitfield) [inline design/parameters.md]: Minor spelling error `paameters` -> `parameters`

**Review by andy-mishechkin — APPROVED** [review]

**Comment by andy-mishechkin** [comment]
@andywhitfield , may you check and resolve comment in this PR, please ?
Thank you!

**Review by fdeschamps-rms — APPROVED** [review]

**Comment by andy-mishechkin** [comment]
> @felix-deschamps , may you check and resolve comment in this PR, please ?
> Thank you!

Already fixed

**Review by EzzhevNikita — APPROVED** [review]

**Review by andy-mishechkin — APPROVED** [review]

**Review by anatolybolshakov — APPROVED** [review]

File: design/parameters.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
+# Loading parameters from external files
+Currently, parameters passed to templates must be declared inline. Here is an example of how one would implement, for example, an expansion of jobs: 
+
+```yaml
+# groups-template.yaml
```
Comment (anatolybolshakov) [inline design/parameters.md:5]: This snippet does not work for me and throws the error: Mapping values are not allowed in this context (row 15 of this file) - could you please check that all snippets here are valid?

**Review by anatolybolshakov — CHANGES_REQUESTED** [review]
Could you please take a look at comments?

File: design/parameters.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        - script: | 
+            echo "the item payload is: ${{ item.payload }}"
+          displayName: "${{ item.displayName }}"
+```
+
+```yaml
+# azure-pipelines.yaml
+
+stages: 
+- stage: Group-Stage
```
Comment (andy-mishechkin) [inline design/parameters.md:24]: This is a wrong Stage name. I suggest to use `Group_Stage`

File: design/parameters.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+# groups-template.yaml
+parameters:
+    groups: []
+jobs:
+  - ${{ each group in parameters.groups }}:
+    - job: ${{ group.name }}
+      pool:
+        vmImage: 'ubuntu-16.04'
+      steps: 
+      - ${{ each item in group.items }}
```
Comment (andy-mishechkin) [inline design/parameters.md:14]: error - you should use - `${{ each item in group.items }}:`

File: design/parameters.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+# azure-pipelines.yaml
+
+stages: 
+- stage: Group-Stage
+  jobs:
+  - template: groups-template.yaml
+    parameters:
+      groups:
+      - group1:
+        name: "first group"
```
Comment (andy-mishechkin) [inline design/parameters.md:30]: Spaces are not allowed in parameter values

File: design/parameters.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        - item: 
+          name: "item1"
+          displayName: "group1-item2"
+          payload: "the first item in group1"
+        - item: 
+          name: "item2"
+          displayName: "group1-item2"
+          payload: "the second item in group1"
+      - group2:
+        name: "second group"
```
Comment (andy-mishechkin) [inline design/parameters.md:41]: Spaces are not allowed in parameter values

File: design/parameters.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+stages: 
+- stage: Group-Stage
+  jobs:
+  - template: groups-template.yaml
+    parameters: 
+    - template: group-data.yaml
+  - template: another-groups-template.yaml
+    parameters: 
+    - template: group-data.yaml
```
Comment (andy-mishechkin) [inline design/parameters.md:96]: This is not work. Please check and provide the working yaml template

**Review by andy-mishechkin — CHANGES_REQUESTED** [review]

**Comment by andy-mishechkin** [comment]
Hello @felix-deschamps 
I've found the issues in your yaml templates (see above). Please check it and test your yaml files in real Azure DevOps project.
Thank you!

**Comment by andy-mishechkin** [comment]
We have to reject this PR, because this approach for using templates for parameters definition doesn't work

**Pull request closed without merging**